include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,839 @@
+        {
+            clustering.erase(c);
+        }
+
+        derive_children_recv_bounds_without_siblings(min_parent);
+
+        construct_merged_charge_state_space(min_parent);
+
+        if (min_parent->charge_space.empty())
+        {
+            terminate = true;
+            return;
+        }
+
+        construct_merged_potential_projections(min_parent);
+
+        compute_meets_for_internal_pot_bounds(min_parent);
+
+        clustering.emplace(min_parent);
+
+        update_charge_spaces(min_parent->uid);
+    }
+
+    [[nodiscard]] constexpr inline uint64_t maximum_top_level_multisets(const uint64_t number_of_sidbs) const noexcept
+    {
+        // computes nCr(N + 2, 2)                             // computes nCr(N + 1, 1)
+        return base == 3 ? ((number_of_sidbs + 1) * (number_of_sidbs + 2)) / 2 : number_of_sidbs + 1;
+    }
+
+    const uint8_t base;
+
+    const sidb_cluster_ptr top_cluster;
+
+    sidb_clustering clustering{};
+
+    bool terminate = false;
+
+    const uint64_t witness_partitioning_max_cluster_size;
+
+    const std::array<double, 4> mu_bounds_with_error;
+};
+
+}  // namespace detail
+
+/**
+ * The purely constructive *Ground State Space* algorithm is the key ingredient of the *ClusterComplete* exact SiDB
+ * simulator that revolutionizes the domain of exact SiDB simulation. It uses iterative "loop until fixpoint" concepts
+ * to prune the simulation search space for not only a flat layout of SiDBs, but rather generalizes, and lifts Jan
+ * Drewniok's game-changing physically informed space pruning technique introduced with *QuickExact* to higher order,
+ * allowing *Ground State Space* to prune multiset charge state configurations at any level in a cluster hierarchy.
+ *
+ * Starting at a clustering of all singleton clusters, the charge spaces, ie. a set of multiset charge configurations
+ * (initially { {{-}}, {{0}}, {{+}} } or omitting the singleton multiset {{+}} in the case of base 2 pre-simulation),
+ * are pruned iteratively through potential bound analysis. Through merges, ie., replacing a set of children in the
+ * clustering with their parent, we may inspect the most crucially dependant interactions in the layout separately. This
+ * is an effect of the defaulted minimal variance cluster linking heuristic, which clusters precisely that which is most
+ * prominently related. The procedure finishes when the charge spaces have been folded all the way up to the top
+ * cluster, parent of all, which then contains all information resulting from the construction. *ClusterComplete*,
+ * without much trickery, now simply unfolds this result, allowing us to simulate problems that were previously seen as
+ * astronomical in size.
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ * @param lyt Layout to construct the ground state space of.
+ * @param max_cluster_size_for_witness_partitioning This specifies the maximum cluster size for which *Ground State
```

Comment:
@wlambooy Is there a reason why it is inilized with ``6``? Can we mention it in the docstring?

Reply:
Empirally. I found that it sometimes gives a little edge without hindering runtimes. Could be set a little higher maybe even. I'll leave that to some future investigation. I'll put a note